Introduce and extend Config pattern (III/2)

sazzad16 opened the pull request:
Resolves #1780 
Resolves #1824 
Resolves #2234 

Closes #1787 
Closes #2237 

Closes #2342 
Closes #2349 
Closes #2352 
Closes #2353 
Closes #2363

mina-asham commented on src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:
```diff
@@ -1,80 +1,122 @@
-    this.sslSocketFactory = sslSocketFactory;
-    this.sslParameters = sslParameters;
-    this.hostnameVerifier = hostnameVerifier;
+    setHost(host);
+    setPort(port);
+    setConnectionTimeout(connectionTimeout);
+    setSoTimeout(soTimeout);
+    setSsl(ssl);
+    setSslSocketFactory(sslSocketFactory);
+    setSslParameters(sslParameters);
+    setHostnameVerifier(hostnameVerifier);
+  }
+
+  public DefaultJedisSocketFactory(HostAndPort hostAndPort, JedisClientConfig config) {
+    setHostAndPort(hostAndPort);
+    if (config != null) {
+      setConnectionTimeout(config.getConnectionTimeout());
+      setSoTimeout(config.getSoTimeout());
+      setSsl(config.isSsl());
+      setSslSocketFactory(config.getSslSocketFactory());
+      setSslParameters(config.getSslParameters());
+      setHostnameVerifier(config.getHostnameVerifier());
+      setHostAndPortMapper(config.getHostAndPortMapper());
```
It's a bit unconventional to use setters in constructors, any advantage of doing that?

sazzad16 replied:
Not exactly. I saw this style is being used in commons-pool2. So I put it here at least to have some discussions :)

mina-asham replied:
IMO, still unconventional commons-pool2 doesn't make it okay :P

mina-asham commented on src/test/java/redis/clients/jedis/tests/JedisClusterTest.java:
```diff
@@ -510,24 +566,35 @@ public void testCloseable() throws IOException {
-        // ok to go...
-      }
-    }
+//    Iterator<JedisPool> poolIterator = jc.getClusterNodes().values().iterator();
+//    while (poolIterator.hasNext()) {
+//      JedisPool pool = poolIterator.next();
+//      try {
+//        pool.getResource();
+//        fail("JedisCluster's internal pools should be already destroyed");
+//      } catch (JedisConnectionException e) {
+//        // ok to go...
+//      }
+//    }
```
I think you forgot to uncomment this?

sazzad16 replied:
Nope.

I debugged and found that `jc.getClusterNodes().values()` is actually empty, even without my changes. So above codes are meaningless.

Also check that following `assertEquals(0, jc.getClusterNodes().size());` is passing.

mina-asham replied:
If that part is useless let's delete it then, commenting it out leave unanswered questions

mina-asham commented on src/test/java/redis/clients/jedis/tests/JedisTest.java:
```diff
@@ -190,14 +179,40 @@ public void allowUrlWithNoDBAndNoPassword() {
 
+  @Test
+  public void uriWithDBindexShouldUseTimeout() throws URISyntaxException {
+    URI uri = new URI("redis://fakehost:6378/1");
+    long startTime = System.nanoTime();
+    try (Jedis j = new Jedis(uri, 5000)) {
+      j.ping();
+    } catch (Exception ex) {
+      assertEquals(JedisConnectionException.class, ex.getClass());
+      assertEquals(java.net.UnknownHostException.class, ex.getCause().getClass());
+    }
+    long stopTime = System.nanoTime();
+    assertTrue(stopTime - startTime > 4000);
```
I don't get this one, the timeout is 5 seconds, but we are validating the runtime is greater than 4 **milli**seconds?

sazzad16 replied:
Socket connections are dependent on OS. We can't strictly check the socket timeouts. I'm using the threshold to reduce this test's probability to become another [flaky](https://github.com/redis/jedis/issues/2367) test.

mina-asham replied:
Got you, so that would make it "less" flaky, but potentially still flaky, do we want to avoid that check completely? We are kind of checking the JDK's Socket timeout works as expected here which is not really adding much, I think just checking the exception is enough in that case

sazzad16 replied:
I really want to have test for #1824 as this is a reported issue. And I can't find any other way to test it.

mina-asham replied:
This should do the trick? (a bit more involved, but should be a lot more stable)

```suggestion
    int port = 6677;
    int timeoutMillis = 5000;
    int deltaMillis = 500;

    URI uri = new URI("redis://localhost:" + port + "/1");
    ServerSocket server = new ServerSocket(port);
    Instant start = Instant.now();
    try (Jedis jedis = new Jedis(uri, timeoutMillis)) {
        jedis.ping();
    } catch (JedisConnectionException ex) {
        assertEquals(SocketTimeoutException.class, ex.getCause().getClass());
        assertEquals(timeoutMillis, Duration.between(start, Instant.now()).toMillis(), deltaMillis);
        return;
    } finally {
        server.close();
    }
    fail();
```

sazzad16 replied:
I'll give that a try.

mina-asham commented on src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:
```diff
@@ -136,6 +137,7 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
+  @Ignore
```
Why ignore it? If the test doesn't make sense anymore can we remove it instead?

sazzad16 replied:
I couldn't make this work with my current changes. What if someone else can? So I kept this.

mina-asham replied:
If it worked before your change then there are two cases:
1. We are testing the wrong thing, in that case let's drop the test
2. We broke something so w need to look into it

I think in either cases we need to look into this one and the other one below

mina-asham commented on src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:
```diff
@@ -145,6 +146,7 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
+  @Ignore
```
Ditto

sazzad16 replied:
Ditto

mina-asham reviewed (changes requested):
Looks good, thanks for going through the comments from the prevision revision! Just a few more minor comments here and then I am happy to sign off

walles commented on src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java:
```diff
@@ -0,0 +1,201 @@
+
+public final class DefaultJedisClientConfig implements JedisClientConfig {
+
+  private final int connectionTimeout;
+  private final int soTimeout;
+  private final int infiniteSoTimeout;
```
With Java 8 (#2357) `Duration` could be used rather than `int`s for the timeouts, which would probably be a better API in the long run.

walles replied:
I assume this goes for all places where `Protocol.DEFAULT_TIMEOUT` is used really, as it is now the unit for the timeouts is not obvious.

sazzad16 replied:
@walles Let's merge this PR first and then we'll think about it :)

gkorland commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -59,16 +65,28 @@ public BinaryClient(final String host, final int port) {
   }
 
+  /**
+   * @deprecated This constructor will be removed in future.
```
@sazzad16 please point to the alternative Constructor

gkorland commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -59,16 +65,28 @@ public BinaryClient(final String host, final int port) {
   }
 
+  /**
+   * @deprecated This constructor will be removed in future.
```
@sazzad16 please point to the alternative Constructor

gkorland commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -81,23 +99,20 @@ public boolean isInWatch() {
-    return result;
+  /**
+   * @param user
+   * @deprecated This method will be removed in future.
```
@sazzad16 please point to the alternative Constructor

sazzad16 replied:
@gkorland There won't be any.

sazzad16 replied:
@gkorland There won't be any direct alternate. Other than config class and, may be, factory class, no other class will hold user data.

gkorland replied:
So you should state in the JDoc

gkorland commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -81,23 +99,20 @@ public boolean isInWatch() {
-
+  /**
+   * @param password
+   * @deprecated This method will be removed in future.
```
@sazzad16 please point to the alternative Constructor

sazzad16 replied:
@gkorland There won't be any direct alternate.

sazzad16 replied:
@gkorland There won't be any direct alternate. Other than config class and, may be, factory class, no other class will hold user data.

gkorland replied:
So you should state it in the JDoc

gkorland commented on src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:
```diff
@@ -136,49 +137,6 @@ public void checkFailedJedisServer() {
   }
 
-  @Test
-  public void shouldReturnActiveShardsWhenOneGoesOffline() {
```
@sazzad16 why did your remove this test?

sazzad16 replied:
@gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

gkorland commented on src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:
```diff
@@ -145,49 +146,6 @@ public void checkFailedJedisServer() {
   }
 
-  @Test
-  public void shouldReturnActiveShardsWhenOneGoesOffline() {
```
@sazzad16 why did your remove this test?

sazzad16 replied:
@gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

gkorland commented on src/test/java/redis/clients/jedis/tests/ConnectionTest.java:
```diff
@@ -4,51 +4,60 @@
   @After
   public void tearDown() throws Exception {
-    client.close();
+    if (client != null) {
```
@sazzad16 why do you still need this `@After` if the connection is not created in the `@Before` any more?

sazzad16 replied:
@gkorland This test class mostly consists of constructor tests. So connections are supposed to be created in each test methods. I'm closing those objects here in `@After`.

gkorland commented on src/main/java/redis/clients/jedis/JedisSocketFactory.java:
```diff
@@ -15,23 +16,29 @@
  */
 public interface JedisSocketFactory {
```
@sazzad16 it seems like all the methods are deprecated so how come the interface isn't?

sazzad16 replied:
@gkorland `createSocket()` method is here to stay. But instead of throwing IOException, it will throw only JedisConnectionException.

gkorland replied:
Yes, since you are not going to really deprecate it

gkorland commented on src/test/java/redis/clients/jedis/tests/ShardedJedisTest.java:
```diff
@@ -270,9 +270,9 @@ public void testMasterSlaveShardingConsistency() {
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT));
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT + 1));
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT + 2));
+    otherShards.add(new JedisShardInfo("127.0.0.1", Protocol.DEFAULT_PORT));
```
why did you change all the host names to IPs?

sazzad16 replied:
@gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

sazzad16 commented:
@gkorland 

1. ShardedJedisPoolTest.shouldReturnActiveShardsWhenOneGoesOffline()
2. ShardedJedisPoolWithCompleteCredentialsTest.shouldReturnActiveShardsWhenOneGoesOffline()
3. ShardedJedisTest.testMasterSlaveShardingConsistency()
4. ShardedJedisTest.testMasterSlaveShardingConsistencyWithShardNaming()

These tests used to work because underlying connections were not always created given the combination of parameters. All the Jedis classes were suffering from this situation. The underlying connections of objects representing **a connection** either should always be created from constructors or should NEVER be created. In this PR I implemented the former approach.

I was able to keep the tests with a valid host address in (3) and (4).

But I could not make (1) and (2) to work. Even though I `@Ignore`d first, I removed those after review. A portion of (1) and (2):
```
    shards.set(1, new JedisShardInfo("localhost", 1234));
    pool = new ShardedJedisPool(redisConfig, shards);
    jedis = pool.getResource();
```
This is more logical that `pool.getResource()` would fail because `1234` is an invalid port. But the tests are expecting `pool.getResource()` without any error. That's why these two tests won't work with this PR.

IMHO, we shouldn't care about impractical tests. Still, I can `@Ignore` those if you prefer.
